Ports bay's supermatter monitor modular program

Shadowlight213 opened the pull request:
Holy shit, actually new content for modular computers instead of refactors!!! :tada: 

:cl:
add: There is now a new monitor program that engineers can use to monitor the supermatter status
/:cl:
Screenshot of the UI (the double colon thing was fixed. this one is a bit old)
![](http://puu.sh/wpb94/edffcaf547.png)

Screen overlays! Laptops use the Eris static version for now. The others will change depending on supermatter status.
![](https://puu.sh/wv5FQ/04dd6152b8.png)

Shadowlight213 committed:
WIP

Bawhoppen commented:
is this able to dynamically monitor any supermatter or just the roundstart engineering ones?

RemieRichards commented on tgui/src/interfaces/ntos_supermatter_monitor.ract:
```diff
@@ -0,0 +1,70 @@
+      }
+    }
+
+  component.exports = {
```
you don't need a second set of ```component.exports``` and ```data```, just put ```BigState(val, aver, max)``` below ```integState```

Like I did here: https://github.com/tgstation/tgstation/blob/eace4eec148e8a366ee2fe6673f852b59a2e77b1/tgui/src/interfaces/mech_bay_power_console.ract#L2

Shadowlight213 replied:
I did that at first, but it gave me compile errors.

Shadowlight213 replied:
@RemieRichards for future reference, is there a way to do an else if, or switch statement in tgui?

RemieRichards replied:
in the template or in the JS at the top of templates?

Shadowlight213 replied:
In the template

RemieRichards replied:
```
{{#if data.example}}
{{else}}
    {{#if data.example_two}}
    {{/if}}
{{/if}}
```
You have to remember that ```else if``` is really just sugar for ```if {} else { if { } }```

Shadowlight213 commented:
Any supermatter. The program starts you on a list of all supermatters detected. It will show any that are on the station or mining level, or on the same level as the user. You can then select one to view in more detail.

Shadowlight213 committed:
How the hell did these not get committed

Bawhoppen commented:
It'd be better if you had to stick a monitoring chip on or something probably instead of just showing all

Shadowlight213 commented:
All supermatters have a monitoring chip on them though. The radio. IC wise, this just interfaces with those existing sensors.

Bawhoppen commented:
I guess

kevinz000 commented:
>doing network refactor
>see this
>SCREAM

meeeeme commented:
neat

Anonmare commented:
Is MeV/cm^3 charge level?

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into smmonitor

Shadowlight213 commented:
yes

Shadowlight213 committed:
Fixes, icons for the tablet

Shadowlight213 commented:
Ok, so I was able to codersprite the overlays for the program onto the tablet, but there is no way I'm managing to get them to fit onto the Eris laptop sprites. The laptop screens are just too small.

Shadowlight213 commented:
Kor won't let me use the bay laptops because I said mean words to oranges, and I cant fit the overlays on the eris ones.

KorPhaeron commented:
????

KorPhaeron commented:
Just use the new sprites you meme

Shadowlight213 commented:
They don't fit. The overlay is too big senpai. It's not going to fit.

KorPhaeron commented:
Then use the bay laptops nobody cares what the laptop sprites are I promise I thought asking permission about that was a joke

optimumtact commented:
maybe try the tip of the overlay?

optimumtact commented:
just the tip

lzimann commented:
@Shadowlight213

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into smmonitor

Shadowlight213 committed:
Laptops use the static engine overlay when running the program.
Engineering tablets now start preloaded with a supermatter monitor and alarm monitor.

Shadowlight213 commented:
Ok, I have the overlay be overridden if a laptop is running the program, as I'd need more time to figure out how to port the bay overlays there.

Shadowlight213 commented:
Whoops.

Shadowlight213 commented:
This is basically done.

Cyberboss commented on code/modules/modular_computers/computers/item/tablet_presets.dm:
```diff
@@ -19,6 +22,12 @@
+/obj/item/device/modular_computer/tablet/preset/advanced/engi/install_programs()
+	var/obj/item/weapon/computer_hardware/hard_drive/hard_drive = all_components[MC_HDD]
+	hard_drive.store_file(new/datum/computer_file/program/alarm_monitor())
+	hard_drive.store_file(new/datum/computer_file/program/supermatter_monitor())
```
Can't say I'm a fan of pre-installing programs, tbh

Shadowlight213 replied:
I can remove it if you want.

Cyberboss replied:
Please do, I believe tablets are eventually supposed to be a: build your own pda sort of deal

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+	ui_header = "smmon_0.gif"
+	program_icon_state = "smmon_0"
+	extended_desc = "This program connects to specially calibrated supermatter sensors to provide information on the status of supermatter-based engines."
+	requires_ntnet = 1
```
TRUE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+	ui_y = 400
+	var/last_status = 0
+	var/list/supermatters
+	var/obj/machinery/power/supermatter_shard/active = null		// Currently selected supermatter crystal.
```
Remove ` = null`

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+			refresh()
+			return
+		var/datum/gas_mixture/air = T.return_air()
+		if(!istype(air))
```
just !air

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+			active = null
+			return
+
+		data["active"] = 1
```
TRUE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+		data["SM_ambientpressure"] = air.return_pressure()
+		//data["SM_EPR"] = round((air.total_moles / air.group_multiplier) / 23.1, 0.01)
+		var/list/gasdata = list()
+		var/list/relevantgas = list("o2","co2","n2","plasma","n2o","freon")
```
why exclude anything? this is basically all gases

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+				"uid" = S.uid
+				)))
+
+		data["active"] = 0
```
FALSE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+
+/datum/computer_file/program/supermatter_monitor/ui_act(action, params)
+	if(..())
+		return 1
```
TRUE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+	switch(action)
+		if("PRG_clear")
+			active = null
+			return 1
```
TRUE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+			return 1
+		if("PRG_refresh")
+			refresh()
+			return 1
```
TRUE

Cyberboss commented on code/modules/modular_computers/file_system/programs/sm_monitor.dm:
```diff
@@ -0,0 +1,136 @@
+			for(var/obj/machinery/power/supermatter_shard/S in supermatters)
+				if(S.uid == newuid)
+					active = S
+			return 1
```
TRUE

Cyberboss commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -165,6 +165,41 @@
+	if(!air)
+		return SUPERMATTER_ERROR
+
+	if(get_integrity() < 5)
```
There should be defines for these levels

Shadowlight213 replied:
You sure? Its a percentage of the sm integrity remaining.

Cyberboss replied:
You're right, Can you make them defines please

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into smmonitor

Shadowlight213 committed:
Defines

Shadowlight213 committed:
Requested changes.
Engineering tablets no longer have preinstalled programs.
Supermatter status percents are defines

Cyberboss committed:
Minor thing
